src/VisualStudio/Core/Def/Implementation/Diagnostics/DiagnosticProgressReporter.cs

```diff
@@ -21,180 +21,168 @@ internal sealed class TaskCenterSolutionAnalysisProgressReporter
         private readonly IVsTaskStatusCenterService _taskCenterService;
         private readonly TaskHandlerOptions _options;
 
-        // these fields are never accessed concurrently
-        private TaskCompletionSource<VoidResult> _currentTask;
-        private DateTimeOffset _lastTimeReported;
-
+        /// <summary>
+        /// Task handler to provide a task to the <see cref="_taskCenterService"/>
+        /// </summary>
+        private ITaskHandler? _taskHandler;
+
+        /// <summary>
+        /// Stores the currently running task center task.
+        /// This is manually started and completed based on receiving start / stop events
+        /// from the <see cref="ISolutionCrawlerProgressReporter"/>
+        /// </summary>
+        private TaskCompletionSource<VoidResult>? _taskCenterTask;
+
+        /// <summary>
+        /// Task used to trigger throttled UI updates in an interval
+        /// defined by <see cref="s_minimumInterval"/>
+        /// </summary>
+        private Task? _intervalTask;
+
+        /// <summary>
+        /// Task used to ensure serialization of UI updates.
+        /// </summary>
+        private Task _updateUITask = Task.CompletedTask;
+
+        /// <summary>
+        /// Gate access to reporting sln crawler events so we cannot
+        /// report UI changes concurrently.
+        /// </summary>
+        private readonly object _lock = new object();
+
+        /// <summary>
+        /// Stores the last shown <see cref="ProgressData"/>
+        /// </summary>
+        private ProgressData _lastProgressData;
+
+        /// <summary>
+        /// Unfortunately, <see cref="ProgressData.PendingItemCount"/> is only reported
+        /// when the <see cref="ProgressData.Status"/> is <see cref="ProgressStatus.PendingItemCountUpdated"/>
+        /// So we have to store the count separately for the UI so that we do not overwrite the last reported count with 0.
+        /// </summary>
         private int _lastPendingItemCount;
```

Comment:
overall this all looks great.  What is missing here are explanations about how these specific pieces of data are protected by the lock/_updateUITask 

I would actually separate them into two sections in the code.  Have the section with the lock and what hte lock protects, then have the section with the _updateUITask  and waht it protects.

Reply:
Like a #region ?  Or perhaps a separate partial file impl, one for throttle, one for UI updating

Reply:
Both of those work for me!  I think in this case a region is totally fine.  It might be a little too much to need a whole separate file.  Your call though!